src/sentry/tasks/releasemonitor.py

```diff
@@ -0,0 +1,165 @@
+                    #       SELECT
+                    #         U0."id"
+                    #       FROM
+                    #         "sentry_release" U0
+                    #       WHERE
+                    #         (
+                    #           U0."organization_id" = 4
+                    #           AND U0."version" = foo @ 1.0.0
+                    #         )
+                    #       LIMIT
+                    #         1
+                    #     )
+                    #     AND "sentry_releaseprojectenvironment"."environment_id" = (
+                    #       SELECT
+                    #         U0."id"
+                    #       FROM
+                    #         "sentry_environment" U0
+                    #       WHERE
+                    #         (
+                    #           U0."organization_id" = 4
+                    #           AND U0."project_id" = 10
+                    #           AND U0."name" = canary
+                    #         )
+                    #       LIMIT
+                    #         1
+                    #     )
+                    #   )
+                    rpe = ReleaseProjectEnvironment.objects.get(
+                        project_id=row["project_id"],
+                        release_id=Subquery(
+                            Release.objects.filter(
+                                organization=org_id, version=row["release"]
+                            ).values("id")[:1],
+                        ),
+                        environment_id=Subquery(
+                            Environment.objects.filter(
+                                organization_id=org_id,
+                                project_id=row["project_id"],
+                                name=row["environment"],
+                            ).values("id")[:1]
+                        ),
+                    )
+                    adopted = (
+                        True
+                        if row["sessions"] / total_sessions >= REQUIRED_ADOPTION_PERCENT
+                        else False
+                    )
+                    if adopted and rpe.adopted is None:
+                        rpe.update(adopted=timezone.now())
+                        updated_ids.append(rpe.id)
+                    elif not adopted and rpe.adopted:
+                        rpe.update(unadopted=timezone.now())
+
+            if not more_results:
+                break
+
+    # 4. Cleanup - releases that are marked as adopted that didn’t get any results in #2 need to be marked as unadopted
+    # Note: I'm really worried here that we would mark all releases as unadopted if snuba failed to return results for any reason...
+    # It would probably get fixed in the next run of the task at least
+    ReleaseProjectEnvironment.objects.filter(unadopted=None).exclude(
+        adopted=None, id__in=updated_ids
+    ).update(unadopted=timezone.now())
```

Comment:
We could possibly handle this by having a counter. So if we see session data below adoption % for release X number of hours in a row then we mark unadopted. Could just be something we store in redis with a ttl.